src/fontlibc/fontlibc.h

```diff
@@ -0,0 +1,437 @@
+
+/**
+ * Allows you to set the code point that is recognized as being a new line code.
+ * You can set this to zero to prevent new line code processing.
+ * This defaults to 0x0A (ASCII line feed/UNIX newline)
+ * @param code_point New code point to use for newline
+ */
+void fontlib_SetNewlineCode(char code_point);
+
+/**
+ * Returns the code point that is currently recognized as being a newline.
+ * @return Current newline
+ */
+char fontlib_GetNewlineCode(void);
+
+/**
+ * Sets an alternate code point to recognize as a stop code.
+ * For example, you can set this to space to make DrawString and GetStringWidth
+ * stop processing when they reach a space.
+ * Setting this to zero will cause this to be ignored.
+ * NULL (0) will still be recognized as a stop code regardless of value.
+ * Defaults to 0.
+ * @param code_point Additional code point to recognize as a stop code.
+ */
+void fontlib_SetAlternateStopCode(char code_point);
+
+/**
+ * Returns the current alternate stop code.
+ * @return Current alternate stop code.
+ */
+char fontlib_GetAlternateStopCode(void);
+
+/**
+ * Sets the first code point considered printable.
+ * All code points before this will be considered control codes.
+ * Setting this to 0 (NULL) will NOT cause NULL to be ignored.
+ * This defaults 0x10.
+ * @param code_point First printable code point
+ */
+void fontlib_SetFirstPrintableCodePoint(char code_point);
+
+/**
+ * Returns the first code point considered printable.
+ * @return Current first printable code point
+ */
+char fontlib_GetFirstPrintableCodePoint(void);
+
+/**
+ * Returns the width of the given glyph
+ * @param codepoint Codepoint to test
+ * @return Width of glyph, 0 if invalid codepoint
+ */
+uint8_t fontlib_GetGlyphWidth(char codepoint);
+
+/**
+ * Returns the width of a string printed in the current font.
+ * Stops processing when it encounters ANY control code or a codepoint not in
+ * the current font.
+ * @param str Pointer to string
+ * @return Width of string
+ */
+uint24_t fontlib_GetStringWidth(char* str);
+
+/**
+ * Returns the width of a string printed in the current font.
+ * Stops processing when it encounters ANY control code or a codepoint not in
+ * the current font, or when max_characters have been processed.
+ * @param str Pointer to string
+ * @param max_characters Maximum number of characters to process
+ * (or 0 if not needed)
+ * @return Width of string
+ */
+uint24_t fontlib_GetStringWidthL(char* str, int24_t max_characters);
+
+/**
+ * Gets the location of the last character processed by GetStringWidth or 
+ * DrawString
+ * @return Pointer to character
+ */
+char* fontlib_GetLastCharacterRead(void);
+
+/**
+ * Returns 0 if DrawStringL or GetStringWidthL returned because max_characters
+ * were processed.  Returns a non-zero number otherwise.
+ * @return Either zero, or a non-zero number depending on the reason DrawStringL
+ * or GetStringWidthL returned.
+ */
+int24_t fontlib_GetCharactersRemaining(void);
```

Comment:
Unless there's a reason why this could be negative,

```suggestion
unsigned fontlib_GetCharactersRemaining(void);
```

Reply:
Yes, it can be negative. PrintString/GetStringWidth internally call PrintStringL/GetStringWidthL but set the maximum length to -1. In that case, the value returned is negative how-many-characters-were-processed.

Reply:
Neither `fontlib_DrawStringL` nor `fontlib_GetStringWidthL` documents the functionality of accepting a negative `max_characters` or returning a negative value.

Reply:
Okay, I see how it works now. But not only is this negative return value case not documented, it's also off by one. You get `-chars_processed - 1`.

Rather than trying to make this fit the narrative of being defined behavior, I'd simply suggest stating that the return value of this function is undefined if the last call to either of these two functions specified no maximum length.

Reply:
I don't want to make it official, so it's not documented. But the idea was that you could also look at how many characters were printed. However, GetLastCharacterRead would be a better choice for figuring that out.

Reply:
Yeah, I think GetLastCharacterRead fills that need fine.

Reply:
I'm suggesting keeping it as signed, but saying that the return value is an undefined but negative value if the last call was to the non-length-limited function variants.

Reply:
I think I mentioned this on IRC without mentioning it anywhere on here, but I realized that the *real* type for all these values that store a number of characters should be `size_t`.